service/notification/NotificationProcessor.php

```diff
@@ -129,11 +129,41 @@ public function processNotification($unprocessedNotification)
 
         switch ($unprocessedNotification['event_code']) {
             case AdyenNotification::AUTHORISATION:
-                $order->setCurrentState(\Configuration::get('PS_OS_PAYMENT'));
-                $this->orderService->addPaymentDataToOrderFromResponse($order, $unprocessedNotification);
+                if ('true' === $unprocessedNotification['success']) {
+                    // AUTHORISATION success = true moves order to paid if order status is not PS_OS_PAYMENT
+                    if ($order->getCurrentState() !== \Configuration::get('PS_OS_PAYMENT')) {
+                        $order->setCurrentState(\Configuration::get('PS_OS_PAYMENT'));
+                    }
+
+                    $this->orderService->addPaymentDataToOrderFromResponse($order, $unprocessedNotification);
+                } else {
+                    if ($order->getCurrentState() !== \Configuration::get('PS_OS_CANCELED')) {
+                        // No previous authorisation notification was processed before
+                        if (!$this->hasProcessedAuthorisationSuccessNotification(
+                            $unprocessedNotification['merchant_reference']
+                        )) {
+                            // AUTHORISATION success = false moves order to canceled if not canceled already
```

Comment:
can we make this an early opt out so we reduct the number of if else statmeents. So something like:
if (!$this->hasProcessedAuthorisationSuccessNotification(
                            $unprocessedNotification['merchant_reference']
return;
                        ))

Reply:
In this exact case I believe an if else makes this scenario easier to understand. Usually I agree that the early returns are better and easier to understand for example when an array is empty then return immediately because there is no use of continue the process (example from this PR: https://github.com/Adyen/adyen-prestashop/pull/84/files/03ac798e276ede922c236a198c6d491bf69a21c4#diff-b28f8e395dca11241416e156fff07337R280-R282)

In this one we have 2 cases which depends on the function: hasProcessedAuthorisationSuccessNotification()

Reply:
It is now really hard to read the code. Can't we include it then in the general if else statements above and add this into the if criteria ?